src/compress.rs

```diff
@@ -0,0 +1,124 @@
+//! Abstract over different compressors that could be used.
+//!
+//! For the time being, we support lz4 and zstd, but since we want a given
+//! shard file to internally specify which compressor was used, we do not
+//! plan to support extensible compressor support. This could be changed in the
+//! future through the use of a compile-time plugin registry system like the
+//! inventory crate.
+
+use std::io::{BufReader, Read, Write};
+
+use anyhow::{bail, Result};
+use zstd::zstd_safe;
+
+/// The compressors supported by shardio.
+#[derive(Clone, Copy)]
+pub enum Compressor {
+    /// lz4 compression at level 2
+    /// this appears to be a good general trad-off of speed and compression ratio.
+    /// note see these benchmarks for useful numers:
+    /// http://quixdb.github.io/squash-benchmark/#ratio-vs-compression
+    /// important note: lz4f (not lz4) is the relevant mode in those charts.
+    Lz4,
+    /// zstd compression using --fast=1
+    /// Roughly the same speed as lz4 level 2 but possibly 40% smaller for certain
+    /// workloads.
+    Zstd,
+}
+
+impl Compressor {
+    const LZ4_ID: &'static [u8; 8] = b"lz4     ";
+    const ZSTD_ID: &'static [u8; 8] = b"zstd    ";
```

Comment:
There are well-established "[magic numbers](https://en.wikipedia.org/wiki/File_format#Magic_number)" for both of these codecs (`28 b5 2f fd` for zstd, `04 22 4d 18` for lz4).  Perhaps we should use them?  And in fact the stream we're encoding will already have them in there, so we don't need to encode this separately.

Reply:
Hmm, that's an interesting idea, I didn't think about relying on those.  I have mixed feelings about that, because that would require additional code to read into a shard, and we'd need to handle weird edge cases.  I think I'll go halfway and switch this existing check over to the current magic numbers, but keep them written here in the index header.

Reply:
What I mean is that you've got a compressed stream, you can defer the decision of which decompressor to use until you start decompressing it, at which point you can instantiate the decompressor based on the observed magic number.